crates/rustc_codegen_spirv/src/linker/destructure_composites.rs

```diff
@@ -0,0 +1,96 @@
+//! will become valid targets for `OpLoad`/`OpStore`.
+use super::apply_rewrite_rules;
+use rspirv::dr::{Instruction, Module, Operand};
+use rspirv::spirv::Op;
+use rustc_data_structures::fx::{FxHashMap, FxHashSet};
+
+pub fn destructure_composites(module: &mut Module) {
+    for function in module.functions.iter_mut() {
+        let mut rewrite_rules = FxHashMap::default();
+        let reference: FxHashMap<_, _> = function
+            .all_inst_iter()
+            .filter_map(|inst| match inst.class.opcode {
+                Op::CompositeConstruct => Some((inst.result_id.unwrap(), inst.clone())),
+                Op::CompositeInsert if inst.operands.len() == 3 => {
+                    Some((inst.result_id.unwrap(), inst.clone()))
+                }
+                _ => None,
+            })
+            .collect();
+        let mut unused: FxHashSet<_> = reference.keys().copied().collect();
+        for inst in function.all_inst_iter_mut() {
+            if inst.class.opcode == Op::CompositeExtract && inst.operands.len() == 2 {
+                let mut composite = inst.operands[0].unwrap_id_ref();
+                let index = inst.operands[1].unwrap_literal_int32();
+
+                let origin = loop {
+                    if let Some(inst) = reference.get(&composite) {
+                        match inst.class.opcode {
+                            Op::CompositeInsert => {
+                                let insert_index = inst.operands[2].unwrap_literal_int32();
+                                if insert_index == index {
+                                    break Some(inst.operands[0].unwrap_id_ref());
+                                }
+                                composite = inst.operands[1].unwrap_id_ref();
+                            }
+                            Op::CompositeConstruct => {
+                                break inst.operands.get(index as usize).map(|o| o.unwrap_id_ref());
+                            }
+                            _ => unreachable!(),
+                        }
+                    } else {
+                        break None;
+                    }
+                };
+
+                if let Some(origin_id) = origin {
+                    rewrite_rules.insert(
+                        inst.result_id.unwrap(),
+                        rewrite_rules.get(&origin_id).map_or(origin_id, |id| *id),
+                    );
+                    *inst = Instruction::new(Op::Nop, None, None, vec![]);
+                    continue;
+                }
+            }
+
+            // If the instruction wasn't replaced, modify the unused set
+            if inst.result_id.is_none() || !reference.contains_key(&inst.result_id.unwrap()) {
+                for op in inst.operands.iter() {
+                    if let Operand::IdRef(id_ref) = op {
+                        unused.remove(id_ref);
+                    }
+                }
+            }
```

Comment:
It's difficult to grok my head around the `unused` logic here (too many double negatives and whatnot, haha) - tbh what might be really nice is to just leave in all the unused stuff as dead code, then let https://github.com/EmbarkStudios/rust-gpu/pull/691 clean it up. That means that this (already kinda complicated and difficult to understand) pass can become simpler, since it doesn't have the responsibility of cleaning up after itself. That's what I did in, for example, mem2reg - it generates too many phis (some of the phis are dead code), but it just leaves that to later optimization passes, since the logic to clean those up already exists in those passes, and there's no need to overcomplicate the mem2reg pass with too many responsibilities.

Reply:
True. This was written before #691, and it's required for correctness to prune those here. Launching a full DCE unconditionally each time we do this opt is way more expensive, meanwhile here we have a lot of info needed for this limited DCE. I'm not convinced.

Reply:
I disagree - we're _already_ "launching a full DCE unconditionally each time we do this opt". We run this opt once, and we do DCE afterwards. It's not more expensive.

Reply:
Fair enough. Blocked by #691 for now, then.